docs/grpc_tutorial.md

```diff
@@ -0,0 +1,265 @@
+Once this is working you can run grpc_examples/run_n_servers.py or run_server.py to run the server(s).
+The script probe_grpc_server will connect to a gRPC server and print the dataloaders and models that are registered
+on it (pass in the address as a parameter).
+The client side of the gRPC communication can then be run using run_grpc_demo.py.
+More details are given below.
+
+## Testing locally with an all-in-one script
+You can test the locally by following the example in mnist_grpc.
+Define your dataloader and model functions as specified above, and register them with the factory.
+Then create n_learners gRPC servers:
+```python
+n_learners = 5
+first_server_port = 9995
+# make n servers
+for i in range(n_learners):
+    port = first_server_port + i
+    server = GRPCServer(mli_factory=ExampleMliFactory(),
+                        port=port)
+    server_process = Process(target=server.run)
+    server_process.start()
+```
+
+And then create n_learners gRPC clients:
+
+```python
+all_learner_models = []
+for i in range(n_learners):
+   port = first_server_port + i
+   ml_system = ExampleGRPCLearnerClient(f"client {i}", f"127.0.0.1:{port}")
+   ml_system.start()
+   dataloader_params = {"train_folder": data_folders[i]}
+   ml_system.setup_ml(dataset_loader_name=dataloader_tag,
+                      dataset_loader_parameters=json.dumps(dataloader_params),
+                      model_arch_name=model_tag,
+                      model_parameters=json.dumps({}))
+   all_learner_models.append(ml_system)
+```
+
+ExampleGRPCLearnerClient implements the machine learning interface, so you can use it with the training functions as before:
+```python
+for round_index in range(n_rounds):
+    results.data.append(
+        collective_learning_round(all_learner_models,
+                                  vote_threshold, round_index)
+    )
+```
+
+## Testing remotely
+We expect that the gRPC learner part will often be on e.g. a compute cluster and be separate from the gRPC client side.
+To test the gRPC in a setup like this you can start the servers on the computer side and the client part separately.
+For one gRPC server:
+```bash
+python3 ./grpc_examples/run_grpc_server.py --port 9995 --metrics_port 9091
+```
+
+For multiple gRPC servers:
+```bash
+python3 ./grpc_examples/run_n_grpc_servers.py --n_learners 5 --port 9995 --metrics_port 9091
+```
+The servers by default will start on port 9995 and use subsequent ports from there, so if three
+servers are required they will run on ports 9995, 9996 and 9997.
+
+If you have written your own dataloaders and models then you need to make sure that those functions are defined or 
+imported before the server is created.
+These are the imports of the default dataloaders and models in `run_grpc_server.py`:
+```python
+# These are imported so that they are registered in the FactoryRegistry
+# pylint: disable=W0611
+import colearn_keras.keras_mnist  # type:ignore # noqa: F401
```

Comment:
Why does the linting need this out of interest?

Reply:
It isn't required for the docs, so I should probably remove it here. In the script it's needed because the imports look like they're unused (but because the imports have the effect of registering the functions with the factory registry they're actually required).